File: radiosim/channel/channel.py
from radiosim.tools.fsm import FSM
from .node_proxy import NodeProxy
import numpy as np
import threading
import queue
import zmq
import logging
log = logging.getLogger(__name__)


class Channel(FSM):
	OFF = "OFF"
	ON = "ON"
	states = [OFF, ON]
	def __init__(self):
		self.register_states( self.states )
		self.initialize( self.OFF )
		self.nodes = {}
		self.ctx = zmq.Context()


	def __repr__(self):
		return f"Channel()"


	def add_node(self, txport=None, rxport=None):
		if txport is None and rxport is None:
			raise ValueError("Must provide one port for either rx or tx")
		self.nodes[f"{txport}{rxport}"] = NodeProxy(channel=self, txport=txport, rxport=rxport)
		log.info(f"Added node to channel: TX-port={txport}, RX-port={rxport}")


	def start(self):
		for key in self.nodes.keys():
			self.nodes[key].start()
		self._core_thread = threading.Thread( target=self._run_core )
		self._core_thread.daemon = True
		self._core_thread.start()


	@FSM.transition(OFF, ON)
	def _run_core(self):
		while self.current_state == self.ON:
			all_signals = []
			for key in self.nodes.keys():
				node = self.nodes[key]
				try:
					sig = node.txbuffer.get( block=False )
				except queue.Empty:
					continue # sig = np.zeros()
				all_signals.append( sig )
			if len(all_signals) == 0: continue
			total_signal = sum(all_signals)
			total_signal += self.generate_noise(len(sig))
			for key in self.nodes.keys():
				node = self.nodes[key]
				node.rxbuffer.put( total_signal )


	def generate_noise(self, num_samples):
		re = np.random.rand(num_samples) 
		im = np.random.rand(num_samples)
		return (re + 1j*im).astype(np.complex64)

File: radiosim/channel/lib/channel.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <chrono>
#include <zmq.hpp>
#include <channel.h>

// Constructor
Channel::Channel(){};


// Destructor
Channel::~Channel(){
	ctx.shutdown();
	ctx.close();
	int num_nodes = nodes.size();
	for (int i=0; i<num_nodes; i++){
		delete nodes[i];
	}
};


// Add a new node to the channel network
void Channel::add_node(unsigned int txport, unsigned int rxport, int buffer_size){
	NodeProxy* new_node = new NodeProxy(ctx, txport, rxport, buffer_size);
	nodes.push_back( new_node );
	std::cout << "Added new node to channel network" << std::endl;
}


// Start the channel
void Channel::start(){
	int num_nodes = nodes.size();
	for (int i=0; i<num_nodes; i++ ){
		std::cout << "Starting node " << i << std::endl;
		nodes[i]->start();
	}
	std::thread loop_thread(&Channel::run_main_loop, this);
	loop_thread.detach();
}


// Runs the main loop of the channel - reading from buffers, processing, and writing to buffers
void Channel::run_main_loop(){
	vector_c64 data;
	data.reserve(512);
	while (true){
		std::this_thread::sleep_for( std::chrono::milliseconds(100) );
		for (auto& node : nodes){
			if (not node->txbuffer.empty()){
				data = node->txbuffer.front();
				node->txbuffer.pop();
				node->rxbuffer.push( data );
			}
		}
	}
}
